### PR TITLE
fix(core): attempt to recover from user errors during creation 

### LINKED ISSUE #31221: Ivy preview report: Error: ASSERTION ERROR: Reached the max number of directives
Hi! Error when click on catalog item... 
Without Ivy all worked correctly.

`Error: ASSERTION ERROR: Reached the max number of directives`

image
image

`<app-store-catalog [baseUrl]="'catalog/'"></app-store-catalog>`
```
<div class="catalog">
  <div *ngFor="let category of (categories$ | async)" [routerLink]="baseUrl + category.path" class="catalog-module">
    <div class="catalog-header">{{category.name}}</div>
    <div class="catalog-body">
      <div class="catalog-img">
        <img [src]="category.image">
      </div>
    </div>
    <div class="catalog-footer">{{category.description}}</div>
  </div>
</div>
```
```
import {Component, Input, OnInit} from '@angular/core';
import {CartHomeService} from "../../../catalog/cart-home.service";
import {map} from "rxjs/operators";

@Component({
  selector: 'app-store-catalog',
  templateUrl: './store-catalog.component.html',
  styleUrls: ['./store-catalog.component.scss']
})
export class StoreCatalogComponent implements OnInit {
  categories$;

  @Input() baseUrl = '';

  constructor(private catalogService: CartHomeService) {
  }

  ngOnInit() {
    this.categories$ = this.catalogService.getList().pipe(
      map(x => x.items)
    );
  }
}
```
```
Angular CLI: 8.1.0-beta.2
Node: 11.14.0
OS: win32 x64
Angular: 8.1.0-next.3
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.801.0-beta.2
@angular-devkit/build-angular      0.801.0-beta.2
@angular-devkit/build-ng-packagr   0.801.0-beta.2
@angular-devkit/build-optimizer    0.801.0-beta.2
@angular-devkit/build-webpack      0.801.0-beta.2
@angular-devkit/core               8.1.0-beta.2
@angular-devkit/schematics         8.1.0-beta.2
@angular/cdk                       8.0.1
@angular/cli                       8.1.0-beta.2
@angular/material                  8.0.1
@ngtools/json-schema               1.1.0
@ngtools/webpack                   8.1.0-beta.2
@schematics/angular                8.1.0-beta.2
@schematics/update                 0.801.0-beta.2
ng-packagr                         5.3.0
rxjs                               6.5.2
typescript                         3.4.5
webpack                            4.34.0
```

Is this a bug?

### PR DESCRIPTION
**Pull request by crisbeto**
If there's an error during the first creation pass of a `TView`, the data structure may be corrupted which will cause framework assertion failures downstream which can mask the user's error. These changes add a new flag to the `TView` that indicates whether the first creation pass was successful, and if it wasn't try re-create the `TView`.

Fixes #31221.

### CODE REVIEW

File: packages/core/src/render3/instructions/shared.ts [hunk 1]
```diff
@@ -373,8 +372,13 @@ export function renderView<T>(tView: TView, lView: LView, context: T): void {
     if (components !== null) {
       renderChildComponents(lView, components);
     }
-
   } finally {
+    // Note that while we set this flag above, we need to do it here as well, because
+    // execution might not reach it if the template function throws an error.
+    if (tView.firstCreatePass) {
+      tView.firstCreatePass = false;
```
Comment (crisbeto) [inline packages/core/src/render3/instructions/shared.ts:379]: Note for reviewer: I was going back-and-forth between doing this and wrapping the template call above with a try/finally. I ended up doing this, because even if we do handle the error above, the query and `renderChildComponents` calls will probably be thrown as well, further masking the user's error.

File: integration/_payload-limits.json [hunk 2]
```diff
@@ -49,7 +49,7 @@
     "master": {
       "uncompressed": {
         "runtime-es2015": 2289,
-        "main-es2015": 226144,
+        "main-es2015": 226674,
```
Comment (crisbeto) [inline integration/_payload-limits.json:52]: I doubt this size increase is from my changes, master is most likely on the edge.

Comment (IgorMinar) [reply integration/_payload-limits.json:52]: yeah, it's unrelated and already being updated in another PR. I can lgtm for this change, but you'll have to wait for AndrewK to review the rest.

**Review by IgorMinar — APPROVED** [review]
lgtm - but only for the unrelated size change

File: packages/core/test/acceptance/view_insertion_spec.ts [hunk 3]
```diff
@@ -661,4 +661,44 @@ describe('view insertion', () => {
     });
 
   });
+
+  it('should not mask user errors during creation with framework errors', () => {
+    // Note that this use case is somewhat contrived, but it's inteded to
+    // mimic the behavior when using the router. See #31221 for more context.
```
Comment (AndrewKushnir) [inline packages/core/test/acceptance/view_insertion_spec.ts:667]: Thanks for adding a comment @crisbeto! The logic in this component is a bit tricky (as you pointed out in the comment), i.e. we have the `<error-wrapper>` in a template and then we also create the same component via `ViewContainerRef.createComponent`, which is also wrapped in try-catch. I think it'd be great to expand the comment a bit to describe why we create the same component and why try-catch block is needed. Also it might make sense to add a reference to a function in Router that performs similar actions.

Otherwise, the changes look good to me 👍

File: packages/core/src/render3/instructions/shared.ts [hunk 4]
```diff
@@ -344,7 +344,6 @@ export function renderView<T>(tView: TView, lView: LView, context: T): void {
     if (templateFn !== null) {
       executeTemplate(tView, lView, templateFn, RenderFlags.Create, context);
     }
-
```
Comment (AndrewKushnir) [inline packages/core/src/render3/instructions/shared.ts:347]: nit: I think having an extra line here is consistent with other blocks below (that are also separated by empty lines), so I'd propose to revert this change.

**Review by AndrewKushnir — APPROVED** [review]
Thanks for fixing this problem and creating a test case for it (which was tricky)! 👍

The changes LGTM (with a couple nits), but I think it'd be great if @pkozlowski-opensource can have a look at this change as well, so I'm adding him as a reviewer to this PR.

Thank you.

**Comment by crisbeto** [comment]
I've addressed the feedback @AndrewKushnir. Also sorry for the larger diff, `clang-format` ended up reformatting the files I changed.

**Comment by crisbeto** [comment]
This is ready for review again.

File: packages/core/src/render3/interfaces/view.ts [hunk 5]
```diff
@@ -684,6 +684,12 @@ export interface TView {
    * Used for directive matching, attribute bindings, local definitions and more.
    */
   consts: TConstants|null;
+
+  /**
+   * Indicates that there was an error before we managed to complete the first create pass of the
+   * view. This means that the view is likely corrupted and we should try to recover it.
+   */
+  incompleteFirstPass: boolean;
```
Comment (crisbeto) [inline packages/core/src/render3/interfaces/view.ts:692]: I was thinking of potentially making this a counter so that we can stop trying after a certain number of attempts. I decided to keep it simple for now, but I'm open to changing it.

Comment (pkozlowski-opensource) [reply packages/core/src/render3/interfaces/view.ts:692]: Yeh, I think that keeping it simple for now is the way to go 👍

File: packages/core/test/acceptance/view_insertion_spec.ts [hunk 6]
```diff
@@ -398,11 +398,11 @@ describe('view insertion', () => {
         const fixture = TestBed.createComponent(AppComponent);
         fixture.detectChanges();
 
-        expect(fixture.debugElement.nativeElement.textContent).toBe('start|test|end');
+        expect(fixture.nativeElement.textContent).toBe('start|test|end');
```
Comment (pkozlowski-opensource) [inline packages/core/test/acceptance/view_insertion_spec.ts:401]: 👍

File: packages/core/test/acceptance/view_insertion_spec.ts [hunk 7]
```diff
@@ -398,11 +398,11 @@ describe('view insertion', () => {
...
 
         fixture.componentInstance.insertTpl = true;
         fixture.detectChanges();
-        expect(fixture.debugElement.nativeElement.textContent).toBe('start|testtest|end');
+        expect(fixture.nativeElement.textContent).toBe('start|testtest|end');
```
Comment (pkozlowski-opensource) [inline packages/core/test/acceptance/view_insertion_spec.ts:405]: 👍

**Review by pkozlowski-opensource — APPROVED** [review]
LGTM.

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:308835838:BASE:308835885:1588090658900:f874594d)

**Comment by AndrewKushnir** [comment]
@crisbeto could you please rebase this PR? There is no conflict, but the merge script failed since this PR was rebased/created prior to the most recent update of the base SHA in the merge script. Feel free to add "merge" label back once it's done.

FYI, g3 presubmit is successful.

**Comment by crisbeto** [comment]
Rebased.

**Comment by AndrewKushnir** [comment]
@crisbeto thanks for rebasing! It looks like payload size limit checks failed on CI, could you please have a look when you get a chance? Thank you.

**Comment by crisbeto** [comment]
Fixed the payload size failures.

**Comment by AndrewKushnir** [comment]
Thanks @crisbeto.

I've re-requested review from @IgorMinar to review/approve payload size increase.

File: goldens/size-tracking/integration-payloads.json [hunk 8]
```diff
@@ -30,7 +30,7 @@
     "master": {
       "uncompressed": {
         "runtime-es2015": 1485,
-        "main-es2015": 137320,
+        "main-es2015": 137897,
```
Comment (kara) [inline goldens/size-tracking/integration-payloads.json:33]: It looks like this is mostly accumulation from other PRs? The change itself doesn't seem to merit this jump in size.

**Review by kara — APPROVED** [review]
LGTM for size changes (look like accumulation)

**Review by IgorMinar — APPROVED** [review]
Lgtm! Thanks!
